Implement filtering and thresholding for plotter

bruno-pannunzio opened the pull request:
This PR proposes the implementation of filtering and thresholding functions from PhasorPy. 

The `phasor_features` table has been updated, adding 2 more columns "G_original" and "S_original", in order to keep the original coordinates so that the filtering and thresholding can be undone if needed. This creates the issue of duplication of information and can be problematic for large images, so we may need to test this.

This features were proposed in #18, and are part of #16.

This PR also fixes some issues with the CalibrationWidget, which had several bugs when calibrating multiple harmonics.

zoccoler commented:
Hi @bruno-pannunzio ,

Thanks for working on this! This feature is really important and useful!

I am testing it on the default synthetic example code defined [here](https://github.com/napari-phasors/napari-phasors/blob/224b9cff79ec5f16673cd7bad3c7378b486a81d3/src/napari_phasors/plotter.py#L619).

After plotting, the threshold and median filter seem to work and I tried changing the slider and the comboboxes and they also work.

But I sometimes get an error after changing the threshold and also after selecting some cluster, check the gif below.

![11 10 2024_16 58 01_REC](https://github.com/user-attachments/assets/29eb1468-e2d1-4055-a2a7-f821205d7850)

Also, here is the error I got:

<details open>
  <summary>Error message</summary>

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_signal.py:1196, in SignalInstance._run_emit_loop(self=<SignalInstance 'color_
indices_changed_signal' on <biaplotter.artists.Histogram2D object>>, args=(<class 'numpy.ndarray'> (1,) int32,))
   1195     with Signal._emitting(self):
-> 1196         self._run_emit_loop_inner()
        self = <SignalInstance 'color_indices_changed_signal' on <biaplotter.artists.Histogram2D object at 0x000001846542E320>>
        self._run_emit_loop_inner = <bound method SignalInstance._run_emit_loop_immediate of <SignalInstance 'color_indices_changed_signal' on
 <biaplotter.artists.Histogram2D object at 0x000001846542E320>>>
   1197 except RecursionError as e:

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_signal.py:1225, in SignalInstance._run_emit_loop_immediate(self=<SignalInstan
ce 'color_indices_changed_signal' on <biaplotter.artists.Histogram2D object>>)
   1224 for caller in self._slots:
-> 1225     caller.cb(args)
        args = (<class 'numpy.ndarray'> (1,) int32,)
        caller = <WeakMethod on napari_phasors.plotter.PlotterWidget.manual_selection_changed>

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_weak_callback.py:453, in WeakMethod.cb(self=<WeakMethod on napari_phasors.plo
tter.PlotterWidget.manual_selection_changed>, args=(<class 'numpy.ndarray'> (1,) int32,))
    452     args = args[: self._max_args]
--> 453 func(obj, *self._args, *args, **self._kwargs)
        obj = <napari_phasors.plotter.PlotterWidget object at 0x000001844B972A70>
        func = <function PlotterWidget.manual_selection_changed at 0x0000018453DAA7A0>
        args = (<class 'numpy.ndarray'> (1,) int32,)
        self = <WeakMethod on napari_phasors.plotter.PlotterWidget.manual_selection_changed>
        self._args = ()
        self._kwargs = {}

File ~\Documents\GitHub\napari-phasors\src\napari_phasors\plotter.py:410, in PlotterWidget.manual_selection_changed(self=<napari_phasors.plott
er.PlotterWidget object>, manual_selection=<class 'numpy.ndarray'> (1,) int32)
    406     harmonic_mask = (
    407         self._labels_layer_with_phasor_features.features["harmonic"]
    408         == harmonic
    409     )
--> 410     self._labels_layer_with_phasor_features.features.loc[
        self._labels_layer_with_phasor_features = <Labels layer 'FLIM data Phasor Features Layer' at 0x184459a0e50>
        self = <napari_phasors.plotter.PlotterWidget object at 0x000001844B972A70>
        column = 'Selection 1'
        harmonic_mask = 0      True
1      True
2      True
3      True
4      True
5      True
6      True
7      True
8      True
9      True
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
Name: harmonic, dtype: bool
        harmonic_mask, column = (0      True
1      True
2      True
3      True
4      True
5      True
6      True
7      True
8      True
9      True
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
Name: harmonic, dtype: bool, 'Selection 1')
        manual_selection = <class 'numpy.ndarray'> (1,) int32
    411         harmonic_mask, column
    412     ] = manual_selection
    413 self.create_phasors_selected_layer()

File ~\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:911, in _LocationIndexer.__setitem__(self=<pandas.core.indexin
g._LocIndexer object>, key=(0      True
1      True
2      True
3      True
...    False
29    False
Name: harmonic, dtype: bool, 'Selection 1'), value=<class 'numpy.ndarray'> (1,) int32)
    910 iloc = self if self.name == "iloc" else self.obj.iloc
--> 911 iloc._setitem_with_indexer(indexer, value, self.name)
        indexer = (<class 'numpy.ndarray'> (30,) bool, 6)
        iloc = <pandas.core.indexing._iLocIndexer object at 0x0000018455AE9BC0>
        self = <pandas.core.indexing._LocIndexer object at 0x0000018455AEAE80>
        value = <class 'numpy.ndarray'> (1,) int32

File ~\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:1942, in _iLocIndexer._setitem_with_indexer(self=<pandas.core.
indexing._iLocIndexer object>, indexer=(<class 'numpy.ndarray'> (30,) bool, 6), value=<class 'numpy.ndarray'> (1,) int32, name='loc')
   1940 if take_split_path:
   1941     # We have to operate column-wise
-> 1942     self._setitem_with_indexer_split_path(indexer, value, name)
        indexer = (<class 'numpy.ndarray'> (30,) bool, 6)
        name = 'loc'
        self = <pandas.core.indexing._iLocIndexer object at 0x0000018455AE9BC0>
        value = <class 'numpy.ndarray'> (1,) int32
   1943 else:

File ~\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:1998, in _iLocIndexer._setitem_with_indexer_split_path(self=<p
andas.core.indexing._iLocIndexer object>, indexer=(<class 'numpy.ndarray'> (30,) bool, 6), value=<class 'numpy.ndarray'> (1,) int32, name='loc
')
   1996         return self._setitem_with_indexer((pi, info_axis[0]), value[0])
-> 1998     raise ValueError(
   1999         "Must have equal len keys and value "
   2000         "when setting with an iterable"
   2001     )
   2003 elif lplane_indexer == 0 and len(value) == len(self.obj.index):
   2004     # We get here in one case via .loc with a all-False mask

ValueError: Must have equal len keys and value when setting with an iterable

The above exception was the direct cause of the following exception:

EmitLoopError                             Traceback (most recent call last)
File ~\Documents\GitHub\napari-phasors\src\napari_phasors\plotter.py:525, in PlotterWidget.plot(self=<napari_phasors.plotter.PlotterWidget obj
ect>)
    521 self.canvas_widget.active_artist = self.canvas_widget.artists[
    522     ArtistType[self.plot_type]
    523 ]
    524 # Set data in the active artist
--> 525 self.canvas_widget.active_artist.data = np.column_stack(
        self.canvas_widget = <biaplotter.plotter.CanvasWidget object at 0x0000018462A12710>
        np.column_stack = <function column_stack at 0x0000018447B64FB0>
        self = <napari_phasors.plotter.PlotterWidget object at 0x000001844B972A70>
        np = <module 'numpy' from 'C:\\Users\\mazo260d\\mambaforge\\envs\\napari-phasors\\lib\\site-packages\\numpy\\__init__.py'>
        x_data = <class 'numpy.ndarray'> (1,) float64
        y_data = <class 'numpy.ndarray'> (1,) float64
        (x_data, y_data) = (<class 'numpy.ndarray'> (1,) float64, <class 'numpy.ndarray'> (1,) float64)
    526     (x_data, y_data)
    527 )
    528 # Set selection data in the active artist
    529 self.canvas_widget.active_artist.color_indices = selection_id_data

File ~\mambaforge\envs\napari-phasors\lib\site-packages\biaplotter\artists.py:333, in Histogram2D.data(self=<biaplotter.artists.Histogram2D ob
ject>, value=<class 'numpy.ndarray'> (1, 2) float64)
    330     if len(color_indices) > color_indices_size:
    331         # fill with zeros where new data is larger
    332         color_indices[color_indices_size:] = 0
--> 333     self.color_indices = color_indices
        self = <biaplotter.artists.Histogram2D object at 0x000001846542E320>
        color_indices = <class 'numpy.ndarray'> (1,) int32
    334 self.draw()

File ~\mambaforge\envs\napari-phasors\lib\site-packages\biaplotter\artists.py:409, in Histogram2D.color_indices(self=<biaplotter.artists.Histo
gram2D object>, indices=<class 'numpy.ndarray'> (1,) int32)
    406 self._overlay = self.ax.imshow(overlay_rgba.swapaxes(0, 1), origin='lower', extent=[
    407     xedges[0], xedges[-1], yedges[0], yedges[-1]], aspect='auto', alpha=1, zorder=2)
    408 # emit signal
--> 409 self.color_indices_changed_signal.emit(self._color_indices)
        self._color_indices = <class 'numpy.ndarray'> (1,) int32
        self = <biaplotter.artists.Histogram2D object at 0x000001846542E320>
        self.color_indices_changed_signal = <SignalInstance 'color_indices_changed_signal' on <biaplotter.artists.Histogram2D object at 0x0000
01846542E320>>
    410 self.draw()

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_signal.py:1176, in SignalInstance.emit(self=<SignalInstance 'color_indices_ch
anged_signal' on <biaplotter.artists.Histogram2D object>>, check_nargs=False, check_types=False, *args=(<class 'numpy.ndarray'> (1,) int32,))
   1172     from ._group import EmissionInfo
   1174     SignalInstance._debug_hook(EmissionInfo(self, args))
-> 1176 self._run_emit_loop(args)
        self = <SignalInstance 'color_indices_changed_signal' on <biaplotter.artists.Histogram2D object at 0x000001846542E320>>
        args = (<class 'numpy.ndarray'> (1,) int32,)

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_signal.py:1213, in SignalInstance._run_emit_loop(self=<SignalInstance 'color_
indices_changed_signal' on <biaplotter.artists.Histogram2D object>>, args=(<class 'numpy.ndarray'> (1,) int32,))
   1205     loop_err = EmitLoopError(
   1206         exc=cb_err,
   1207         signal=self,
   (...)
   1210         emit_queue=self._emit_queue,
   1211     ).with_traceback(cb_err.__traceback__)
   1212     # this comment will show up in the traceback
-> 1213     raise loop_err from cb_err  # emit() call ABOVE || callback error BELOW
        loop_err = EmitLoopError("\n\nWhile emitting signal 'biaplotter.artists.Histogram2D.color_indices_changed_signal', a ValueError occurr
ed in a callback:\n\n  Signal emitted at: C:\\Users\\mazo260d\\mambaforge\\envs\\napari-phasors\\lib\\site-packages\\biaplotter\\artists.py:40
9, in color_indices\n    >  self.color_indices_changed_signal.emit(self._color_indices)\n\n  Callback error at: C:\\Users\\mazo260d\\mambaforg
e\\envs\\napari-phasors\\lib\\site-packages\\pandas\\core\\indexing.py:1998, in _setitem_with_indexer_split_path\n    >  raise ValueError(\n\n
    Local variables:\n       indexer = (array([ True,  True,  True,  True,  True,  True,  True,  Tr...\n       value = array([0])\n       name
 = 'loc'\n       info_axis = 6\n       ilocs = [6]\n       pi = array([ True,  True,  True,  True,  True,  True,  True,  Tru...\n       lplane
_indexer = 10\n\nSee ValueError above for original traceback.")
   1214 finally:
   1215     self._recursion_depth -= 1

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_signal.py:1196, in SignalInstance._run_emit_loop(self=<SignalInstance 'color_
indices_changed_signal' on <biaplotter.artists.Histogram2D object>>, args=(<class 'numpy.ndarray'> (1,) int32,))
   1192     self._max_recursion_depth = max(
   1193         self._max_recursion_depth, self._recursion_depth
   1194     )
   1195     with Signal._emitting(self):
-> 1196         self._run_emit_loop_inner()
        self = <SignalInstance 'color_indices_changed_signal' on <biaplotter.artists.Histogram2D object at 0x000001846542E320>>
        self._run_emit_loop_inner = <bound method SignalInstance._run_emit_loop_immediate of <SignalInstance 'color_indices_changed_signal' on
 <biaplotter.artists.Histogram2D object at 0x000001846542E320>>>
   1197 except RecursionError as e:
   1198     raise RecursionError(
   1199         f"RecursionError when "
   1200         f"emitting signal {self.name!r} with args {args}"
   1201     ) from e

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_signal.py:1225, in SignalInstance._run_emit_loop_immediate(self=<SignalInstan
ce 'color_indices_changed_signal' on <biaplotter.artists.Histogram2D object>>)
   1223 args = self._emit_queue.popleft()
   1224 for caller in self._slots:
-> 1225     caller.cb(args)
        args = (<class 'numpy.ndarray'> (1,) int32,)
        caller = <WeakMethod on napari_phasors.plotter.PlotterWidget.manual_selection_changed>

File ~\mambaforge\envs\napari-phasors\lib\site-packages\psygnal\_weak_callback.py:453, in WeakMethod.cb(self=<WeakMethod on napari_phasors.plo
tter.PlotterWidget.manual_selection_changed>, args=(<class 'numpy.ndarray'> (1,) int32,))
    451 if self._max_args is not None:
    452     args = args[: self._max_args]
--> 453 func(obj, *self._args, *args, **self._kwargs)
        obj = <napari_phasors.plotter.PlotterWidget object at 0x000001844B972A70>
        func = <function PlotterWidget.manual_selection_changed at 0x0000018453DAA7A0>
        args = (<class 'numpy.ndarray'> (1,) int32,)
        self = <WeakMethod on napari_phasors.plotter.PlotterWidget.manual_selection_changed>
        self._args = ()
        self._kwargs = {}

File ~\Documents\GitHub\napari-phasors\src\napari_phasors\plotter.py:410, in PlotterWidget.manual_selection_changed(self=<napari_phasors.plott
er.PlotterWidget object>, manual_selection=<class 'numpy.ndarray'> (1,) int32)
    401 for harmonic in range(
    402     1,
    403     self._labels_layer_with_phasor_features.features["harmonic"].max()
    404     + 1,
    405 ):
    406     harmonic_mask = (
    407         self._labels_layer_with_phasor_features.features["harmonic"]
    408         == harmonic
    409     )
--> 410     self._labels_layer_with_phasor_features.features.loc[
        self._labels_layer_with_phasor_features = <Labels layer 'FLIM data Phasor Features Layer' at 0x184459a0e50>
        self = <napari_phasors.plotter.PlotterWidget object at 0x000001844B972A70>
        column = 'Selection 1'
        harmonic_mask = 0      True
1      True
2      True
3      True
4      True
5      True
6      True
7      True
8      True
9      True
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
Name: harmonic, dtype: bool
        harmonic_mask, column = (0      True
1      True
2      True
3      True
4      True
5      True
6      True
7      True
8      True
9      True
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
Name: harmonic, dtype: bool, 'Selection 1')
        manual_selection = <class 'numpy.ndarray'> (1,) int32
    411         harmonic_mask, column
    412     ] = manual_selection
    413 self.create_phasors_selected_layer()

File ~\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:911, in _LocationIndexer.__setitem__(self=<pandas.core.indexin
g._LocIndexer object>, key=(0      True
1      True
2      True
3      True
...    False
29    False
Name: harmonic, dtype: bool, 'Selection 1'), value=<class 'numpy.ndarray'> (1,) int32)
    908 self._has_valid_setitem_indexer(key)
    910 iloc = self if self.name == "iloc" else self.obj.iloc
--> 911 iloc._setitem_with_indexer(indexer, value, self.name)
        indexer = (<class 'numpy.ndarray'> (30,) bool, 6)
        iloc = <pandas.core.indexing._iLocIndexer object at 0x0000018455AE9BC0>
        self = <pandas.core.indexing._LocIndexer object at 0x0000018455AEAE80>
        value = <class 'numpy.ndarray'> (1,) int32

File ~\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:1942, in _iLocIndexer._setitem_with_indexer(self=<pandas.core.
indexing._iLocIndexer object>, indexer=(<class 'numpy.ndarray'> (30,) bool, 6), value=<class 'numpy.ndarray'> (1,) int32, name='loc')
   1939 # align and set the values
   1940 if take_split_path:
   1941     # We have to operate column-wise
-> 1942     self._setitem_with_indexer_split_path(indexer, value, name)
        indexer = (<class 'numpy.ndarray'> (30,) bool, 6)
        name = 'loc'
        self = <pandas.core.indexing._iLocIndexer object at 0x0000018455AE9BC0>
        value = <class 'numpy.ndarray'> (1,) int32
   1943 else:
   1944     self._setitem_single_block(indexer, value, name)

File ~\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:1998, in _iLocIndexer._setitem_with_indexer_split_path(self=<p
andas.core.indexing._iLocIndexer object>, indexer=(<class 'numpy.ndarray'> (30,) bool, 6), value=<class 'numpy.ndarray'> (1,) int32, name='loc
')
   1993     if len(value) == 1 and not is_integer(info_axis):
   1994         # This is a case like df.iloc[:3, [1]] = [0]
   1995         #  where we treat as df.iloc[:3, 1] = 0
   1996         return self._setitem_with_indexer((pi, info_axis[0]), value[0])
-> 1998     raise ValueError(
   1999         "Must have equal len keys and value "
   2000         "when setting with an iterable"
   2001     )
   2003 elif lplane_indexer == 0 and len(value) == len(self.obj.index):
   2004     # We get here in one case via .loc with a all-False mask
   2005     pass

EmitLoopError:

While emitting signal 'biaplotter.artists.Histogram2D.color_indices_changed_signal', a ValueError occurred in a callback:

  Signal emitted at: C:\Users\mazo260d\mambaforge\envs\napari-phasors\lib\site-packages\biaplotter\artists.py:409, in color_indices
    >  self.color_indices_changed_signal.emit(self._color_indices)

  Callback error at: C:\Users\mazo260d\mambaforge\envs\napari-phasors\lib\site-packages\pandas\core\indexing.py:1998, in _setitem_with_indexer
_split_path
    >  raise ValueError(

    Local variables:
       indexer = (array([ True,  True,  True,  True,  True,  True,  True,  Tr...
       value = array([0])
       name = 'loc'
       info_axis = 6
       ilocs = [6]
       pi = array([ True,  True,  True,  True,  True,  True,  True,  Tru...
       lplane_indexer = 10

See ValueError above for original traceback.
```

</details>

I think it is because the threshold masks/removes some data.

Also, 2 minor things:
- the minimum value for the "Number of time to apply median filter" is set to 1, but I think it should be zero in case the user does not want to apply the filter
- How do you determine the min/max values of the slider? In the example case above, it goes from 0 to 20, but the data max value is ~0.36, which makes the step of the slider very harsh

Let me know what you think of these tests and ideas

bruno-pannunzio commented:
Hi @zoccoler and thanks for all the comments. 

> Also, 2 minor things:
> 
> * the minimum value for the "Number of time to apply median filter" is set to 1, but I think it should be zero in case the user does not want to apply the filter
> * How do you determine the min/max values of the slider? In the example case above, it goes from 0 to 20, but the data max value is ~0.36, which makes the step of the slider very harsh

You are right with these two. I changed to set the number of times to apply the median filter to default 0. Also I modified the threshold slider so that it dynamically changes the range and the steps of each increased based on the maximum value of the mean intensity image.

Regarding the bug:

I modified some functions of the plotter to handle the missing values after thresholding or that were NaN after phasor transform. I had to create a mask to ignore those invalid values from the manual selection. The only drawback is that now if you select some  phasors and then apply a higher threshold and those values are filtered out, then the older selection is 'removed' (set as 0). But then, I can't think of a better solution for this.

zoccoler commented:
It looks great now, thanks! 🎉 

> The only drawback is that now if you select some phasors and then apply a higher threshold and those values are filtered out, then the older selection is 'removed' (set as 0). But then, I can't think of a better solution for this.

That solution currently makes sense and it is easier at the moment than caching selections, so I am in for it!
